conjure-core/src/main/java/com/palantir/conjure/defs/validator/EndpointDefinitionValidator.java

```diff
@@ -99,17 +100,32 @@ public void validate(EndpointDefinition definition, DealiasingTypeVisitor dealia
             definition.getArgs()
                     .stream()
                     .filter(arg -> !arg.getParamType().accept(ParameterTypeVisitor.IS_BODY))
-                    .forEach(arg -> {
-                        boolean isValid = dealiasingTypeVisitor.dealias(arg.getType())
-                                .fold(
-                                        typeDefinition -> true,
-                                        type -> !type.accept(TypeVisitor.IS_BINARY)
-                                                && !type.accept(TypeVisitor.IS_ANY)
-                                );
-                        Preconditions.checkArgument(
-                                isValid, "Non body parameters cannot be of the 'binary' type: '%s' is not allowed",
-                                arg.getArgName());
-                    });
+                    .forEach(arg -> Preconditions.checkArgument(
+                            validateType(arg.getType(), dealiasingTypeVisitor),
+                            "Non body parameters cannot contain the 'binary' type. "
+                                    + "Parameter '%s' from endpoint '%s' violates this constraint.",
+                            arg.getArgName(),
+                            describe(definition)));
+        }
+
+        private static boolean validateType(Type input, DealiasingTypeVisitor dealiasingTypeVisitor) {
+            Optional<Type> dealiased = dealiasingTypeVisitor.dealias(input)
+                    .fold(typeDefinition -> Optional.empty(), Optional::of);
+            // typeDef isn't binary
+            if (!dealiased.isPresent()) {
+                return true;
+            }
+            Type type = dealiased.get();
+            if (input.accept(TypeVisitor.IS_OPTIONAL)) {
```

Comment:
rather than a chain of if statements here, shall we just use a Type.Visitor so that it's guaranteed to be exhaustive?

Reply:
The visitor creates a lot of duplication with the dealiasing visitor and requires us to handle types that are already disallowed elsewhere (map). I think the current code is easier to follow, but it would be more difficult to find if we added new container types.

Reply:
OK fair enough - guess we're not at particular risk from that as no new collection types coming down the pipeline!